src/marlin_pc/constraints.rs

```diff
@@ -0,0 +1,1978 @@
+                proofs,
+                evals,
+                polynomial: PhantomData,
+            }
+        })
+    }
+}
+
+/// Var for the Marlin-KZG10 polynomial commitment verifier.
+pub struct MarlinKZG10Gadget<CycleE, P, PG>
+where
+    CycleE: CycleEngine,
+    P: UVPolynomial<<CycleE::E2 as PairingEngine>::Fr, Point = <CycleE::E2 as PairingEngine>::Fr>,
+    PG: PairingVar<CycleE::E2, <CycleE::E1 as PairingEngine>::Fr>,
+    <CycleE::E2 as PairingEngine>::G1Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G2Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G1Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+    <CycleE::E2 as PairingEngine>::G2Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+{
+    _cycle_engine: PhantomData<CycleE>,
+    _pairing_gadget: PhantomData<PG>,
+    _polynomial: PhantomData<P>,
+}
+
+impl<CycleE, P, PG> Clone for MarlinKZG10Gadget<CycleE, P, PG>
+where
+    CycleE: CycleEngine,
+    P: UVPolynomial<<CycleE::E2 as PairingEngine>::Fr, Point = <CycleE::E2 as PairingEngine>::Fr>,
+    PG: PairingVar<CycleE::E2, <CycleE::E1 as PairingEngine>::Fr>,
+    <CycleE::E2 as PairingEngine>::G1Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G2Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G1Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+    <CycleE::E2 as PairingEngine>::G2Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+{
+    fn clone(&self) -> Self {
+        MarlinKZG10Gadget {
+            _cycle_engine: PhantomData,
+            _pairing_gadget: PhantomData,
+            _polynomial: PhantomData,
+        }
+    }
+}
+
+impl<CycleE, P, PG> MarlinKZG10Gadget<CycleE, P, PG>
+where
+    CycleE: CycleEngine,
+    P: UVPolynomial<<CycleE::E2 as PairingEngine>::Fr, Point = <CycleE::E2 as PairingEngine>::Fr>,
+    for<'a, 'b> &'a P: Div<&'b P, Output = P>,
+    PG: PairingVar<CycleE::E2, <CycleE::E1 as PairingEngine>::Fr>,
+    <CycleE::E2 as PairingEngine>::G1Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G2Projective: MulAssign<<CycleE::E1 as PairingEngine>::Fq>,
+    <CycleE::E2 as PairingEngine>::G1Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+    <CycleE::E2 as PairingEngine>::G2Affine:
+        ToConstraintField<<<CycleE::E1 as PairingEngine>::Fr as Field>::BasePrimeField>,
+{
+    #[allow(clippy::type_complexity, clippy::too_many_arguments)]
+    #[tracing::instrument(
+        target = "r1cs",
+        skip(prepared_verification_key, lc_info, query_set, evaluations, proofs)
+    )]
+    fn prepared_batch_check_evaluations(
```

Comment:
Review needed: This is a very long function. We may not be able to reorganize it this time, but likely we would do something in the future. For example, `lc_info` is too long.